### PR TITLE
gh-108337: Add pyatomic.h header

### PR DESCRIPTION
**Pull request by colesbury**
This adds a new header that provides atomic operations on common data types.

Implementing PEP 703 requires use of atomic operations on more data types than provided by pycore_atomic.h. Additionally, pycore_atomic.h is only usable from Py_BUILD_CORE modules; it can't be used in public headers. PEP 703 will require atomic operations in object.h for Py_INCREF/DECREF, for example. The intention is that this will be exposed through Python.h, although that is not the case yet.

To avoid build issues in third-party extensions, the `pyatomic.h` header generally does not require `-std=gnu11` or `-std=c11` to be passed to the compiler (for GCC or Clang). When compiling C, MSVC will use the `pyatomic_msc.h`, which uses compiler intrinsics. When compiled in C++ mode, MSVC will use the `pyatomic_std.h` implementation, which uses C++11 atomics.

* Issue: gh-108337

### CODE REVIEW

**Comment by vstinner** [comment]
>  it can't be used in public headers

Is this API public or private? Names are prefixed by ``_Py``, so it looks private. I'm trying to *remove* private APIs from the public API, rather than *adding* them :-) Should it be usable by 3rd party projects? Or is it ok to restrict its usage to Python?

At least, I suggest to exclude it from the limited C API for now: so move it to cpython/ directory, and check for Py_LIMITED_API. See other header files like Include/cpython/pydebug.h, the ones included with cpython/ in Python.h.

---

I like the approach using **functions** rather than atomic **types**: it's the approach used by the glib library: https://developer-old.gnome.org/glib/stable/glib-Atomic-Operations.html I'm curious why/how some function parameters don't need volatile, like g_atomic_int_inc():

> While atomic has a volatile qualifier, this is a historical artifact and **the pointer passed to it should not be volatile**.

I wrote Include/internal/pycore_atomic_funcs.h which tries to address Include/internal/pycore_atomic.h compiler issues. But pycore_atomic_funcs.h is incomplete and was only used for a very few things (is it still used?).

---

Using atomic variables is hard :-( Do you have any kind of documentation? Or links to other documentations?

It would be good to have a doc, at least in pyatomic.h.

**Comment by encukou** [comment]
> Is this API public or private? Names are prefixed by _Py, so it looks private. I'm trying to remove private APIs from the public API, rather than adding them :-) Should it be usable by 3rd party projects? Or is it ok to restrict its usage to Python?

They are private, but they're `static inline` functions that will be used by public `static inline` functions/macros (`Py_INCREF`/`DECREF`). So they need to be visible to the compiler without `Py_BUILD_CORE`.
The leading underscore is the way to go here.

Although, `Include/pyatomic.h` should go in `Include/cpython` too -- it's not stable ABI.

> But pycore_atomic_funcs.h is incomplete and was only used for a very few things (is it still used?).

Yes, but AFAICS the uses can be replaced.

**Comment by colesbury** [comment]
As @encukou wrote, the intention is that they're private and only used directly by CPython for now. I'll move them to the `Include/cpython` directory.

Regarding `volatile`, it's only meaningful for the MSVC implementation, which doesn't support C11 atomics yet (*). This can be pushed down into the MSVC function bodies so it's no longer part of the common function signatures.

I'll add some code documentation and links in pyatomic.h

(*) There's experimental support in the most recent builds.

**Comment by vstinner** [comment]
> the intention is that they're private and only used directly by CPython for now

Unless there is a good reason to expose this API to 3rd party code, I would suggest to move it to Include/internal/. For example, override pycore_atomic_funcs.h.

If the header file is only exposed in the internal C API, we will have less compilation issues.

Even after I moved ``Include/pyatomic.h`` to ``Include/internal/pycore_pyatomic.h``, people get compilation errors... Someone wants to use ``pycore_dict.h`` in C++, but ``pycore_dict.h`` indirectly includes ``pycore_pyatomic.h`` which is incompatible with C++ :-(

Is your header file, declaration **and** implementation (since there are static inline functions, the implementation is public!), compatible with C++?

To avoid compilation issues, would it be technically possible to have:

* **internal** C API: fast **static inline** implementation, name prefixed by ``Py_``, implemented in header files.
* **public** C API: slow **regular functions** (opaque function calls, similar to the glib library), name prefixed by ``Py``. Implementation calls the internal C API.

**Comment by encukou** [comment]
Oh, right, atomics are an optional C11 feature!

As far as I can see, the current `pycore_atomic.h` has an unsafe fallback to `volatile`, which is presumably OK-ish for systems/compilers without atomics? That feels a bit cheaty. IMO, we should just explicitly require C11/C++/MSVC atomics.

[PEP-7](https://peps.python.org/pep-0007/#c-dialect) should be updated, and this should get a What's New entry similar to the [3.11 one for "Building CPython"](https://docs.python.org/3.11/whatsnew/3.11.html#build-changes). You might want have PEP-703 state the compiler requirement explicitly.

**Comment by zooba** [comment]
> To avoid compilation issues, would it be technically possible to have:
> 
> * **internal** C API: fast **static inline** implementation, name prefixed by `Py_`, implemented in header files.
> * **public** C API: slow **regular functions** (opaque function calls, similar to the glib library), name prefixed by `Py`. Implementation calls the internal C API.

I agree with this, but I suspect we're going to need these available for macros/inlines. So they'll have to be available in the public API, even if they're not intended for direct use.

Stabe API shouldn't include them, obviously, which means any that's currently a macro/inline for the stable API can't use them either and needs an opaque function call.

**Comment by colesbury** [comment]
@vstinner:
 * The code will be used indirectly in other `static inline` functions that are public so it needs to be visible; it can't be in `Include/internal` because it will need to be used by, for example, the `static inline` Py_INCREF function.
 * This implementation avoids the compilation errors people saw with `pycore_pyatomic.h` because of the preference builtins/intrinsics. (i.e., it doesn't require `-std=c11`)

@encukou: 
* I'll update the what's new, but practically, this does not change compiler requirements from 3.11.
 
@zooba:
* Yeah, as you wrote, we're going to need these available for macros/inlines. I think we can avoid using them in the stable ABI.

**Comment by vstinner** [comment]
> The code will be used indirectly in other static inline functions that are public

Oh ok, now I get it. It wasn't clear when I first reviewed your PR. In that case, it can be in Include/cpython/.

Do you know if you need any atomic function in Include/ header files (limited C API)?

**Comment by colesbury** [comment]
@vstinner:

> Do you know if you need any atomic function in Include/ header files (limited C API)?

In `Include/object.h` but only when `Py_LIMITED_API` is not defined. (For reference counting in `--disable-gil` builds.)

File: Include/cpython/pyatomic.h [hunk 1]
```diff
@@ -0,0 +1,397 @@
+#ifndef Py_ATOMIC_H
+#define Py_ATOMIC_H
+
+// This header provides cross-platform low-level atomic operations
```
Comment (vstinner) [inline Include/cpython/pyatomic.h:1]: You can move this help at the top, before the #ifndef/#define dance.

File: Include/cpython/pyatomic.h [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+
+#ifndef _Py_USE_GCC_BUILTIN_ATOMICS
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 8))
+#define _Py_USE_GCC_BUILTIN_ATOMICS 1
+#elif defined(__clang__)
+#if __has_builtin(__atomic_load)
+#define _Py_USE_GCC_BUILTIN_ATOMICS 1
+#endif
+#endif
+#endif
```
Comment (vstinner) [inline Include/cpython/pyatomic.h]: Indent for readability:

```suggestion
#ifndef _Py_USE_GCC_BUILTIN_ATOMICS
#  if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 8))
#    define _Py_USE_GCC_BUILTIN_ATOMICS 1
#  elif defined(__clang__)
#    if __has_builtin(__atomic_load)
#      define _Py_USE_GCC_BUILTIN_ATOMICS 1
#    endif
#  endif
#endif
```

File: Include/cpython/pyatomic.h [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+#include "cpython/pyatomic_gcc.h"
+#elif __STDC_VERSION__ >= 201112L && !defined(__STDC_NO_ATOMICS__)
+#define Py_ATOMIC_STD_H
+#include "cpython/pyatomic_std.h"
+#elif defined(_MSC_VER)
+#define Py_ATOMIC_MSC_H
+#include "cpython/pyatomic_msc.h"
+#else
+#error "define pyatomic for this platform"
+#endif
```
Comment (vstinner) [inline Include/cpython/pyatomic.h]: Indent + change error message:

```suggestion
#if _Py_USE_GCC_BUILTIN_ATOMICS
#  define Py_ATOMIC_GCC_H
#  include "cpython/pyatomic_gcc.h"
#elif __STDC_VERSION__ >= 201112L && !defined(__STDC_NO_ATOMICS__)
#  define Py_ATOMIC_STD_H
#  include "cpython/pyatomic_std.h"
#elif defined(_MSC_VER)
#  define Py_ATOMIC_MSC_H
#  include "cpython/pyatomic_msc.h"
#else
#  error "no available pyatomic implementation for this platform/compiler"
#endif
```

File: Include/cpython/pyatomic_gcc.h [hunk 4]
```diff
@@ -0,0 +1,679 @@
+#ifndef Py_ATOMIC_GCC_H
+#  error "this header file must not be included directly"
+#endif
+
+// This is the implementation of Python atomic operations using GCC's built-in
```
Comment (vstinner) [inline Include/cpython/pyatomic_gcc.h]: I suggest moving it at the top of the file.

File: Include/cpython/pyatomic_gcc.h [hunk 5]
```diff
@@ -0,0 +1,679 @@
...
+// functions that match the C+11 memory model. This implementation is preferred
+// for GCC compatible compilers, such as Clang. These functions are available in
+// GCC 4.8+ without needing to compile with --std=c11 or --std=gnu11.
+
+static inline int
+_Py_atomic_add_int(int *address, int value)
+{
+    return __atomic_fetch_add(address, value, __ATOMIC_SEQ_CST);
+}
```
Comment (vstinner) [inline Include/cpython/pyatomic_gcc.h]: Usually, I prefer verbose syntax like the one that you used. But in this header file, you have tons of static inline functions, so I suggest using the compact syntax:

```suggestion
static inline int
_Py_atomic_add_int(int *address, int value)
{ return __atomic_fetch_add(address, value, __ATOMIC_SEQ_CST); }
```

File: Include/cpython/pyatomic.h [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+// the previous value of the atomic variable, not the new value.
+//
+// See https://en.cppreference.com/w/c/atomic for more information on C11 atomics.
+// See https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2055r0.pdf
+// "A Relaxed Guide to memory_order_relaxed" for discussion of and common usage
+// or relaxed atomics.
+
+// Atomically adds `value` to `address` and returns the previous value
+static inline int
+_Py_atomic_add_int(int *address, int value);
```
Comment (vstinner) [inline Include/cpython/pyatomic.h]: The glib library uses *atomic* for the first argument. Maybe it's a better name than *address* or *pointer*, what do you think?

File: Include/cpython/pyatomic_gcc.h [hunk 7]
```diff
@@ -0,0 +1,679 @@
...
+_Py_atomic_add_ssize(Py_ssize_t *address, Py_ssize_t value)
+{
+    return __atomic_fetch_add(address, value, __ATOMIC_SEQ_CST);
+}
+
+
+static inline int
+_Py_atomic_compare_exchange_int(int *address, int expected, int value)
+{
+    return __atomic_compare_exchange_n(address, &expected, value, 0, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
```
Comment (vstinner) [inline Include/cpython/pyatomic_gcc.h]: Can you try to git into 80 columns? (Write this call in two lines.) Same remark for code below.

File: Include/cpython/pyatomic_gcc.h [hunk 8]
```diff
@@ -0,0 +1,679 @@
...
+_Py_atomic_compare_exchange_ssize(Py_ssize_t *address, Py_ssize_t expected, Py_ssize_t value)
+{
+    return __atomic_compare_exchange_n(address, &expected, value, 0, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+}
+
+static inline int
+_Py_atomic_compare_exchange_ptr(void *address, void *expected, void *value)
+{
+    void *e = expected;
+    return __atomic_compare_exchange_n((void **)address, &e, value, 0, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
```
Comment (vstinner) [inline Include/cpython/pyatomic_gcc.h]: Wait, that's strange. address type should be ``void**``, no? Can you explain this code?

File: Include/cpython/pyatomic.h [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+
+static inline uintptr_t
+_Py_atomic_add_uintptr(uintptr_t *address, uintptr_t value);
+
+static inline Py_ssize_t
+_Py_atomic_add_ssize(Py_ssize_t *address, Py_ssize_t value);
+
+// Performs an atomic compare-and-exchange. If `*address` and `expected` are equal,
+// then `value` is stored in `*address`. Returns 1 on success and 0 on failure.
+// These correspond to the "strong" variations of the C11 atomic_compare_exchange_* functions.
```
Comment (pitrou) [inline Include/cpython/pyatomic.h]: C11 passes `expected` as a pointer, so that it's updated with the actual value when the latter doesn't match the former. Why not keep that convention here?

File: Include/cpython/pyatomic.h [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+_Py_atomic_add_uint32(uint32_t *address, uint32_t value);
+
+static inline uint64_t
+_Py_atomic_add_uint64(uint64_t *address, uint64_t value);
+
+static inline uintptr_t
+_Py_atomic_add_uintptr(uintptr_t *address, uintptr_t value);
+
+static inline Py_ssize_t
+_Py_atomic_add_ssize(Py_ssize_t *address, Py_ssize_t value);
```
Comment (pitrou) [inline Include/cpython/pyatomic.h]: Not sure it's worth exposing atomic ops for all int sizes and signednesses?

File: Include/cpython/pyatomic.h [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+_Py_atomic_compare_exchange_uint64(uint64_t *address, uint64_t expected, uint64_t value);
+
+static inline int
+_Py_atomic_compare_exchange_uintptr(uintptr_t *address, uintptr_t expected, uintptr_t value);
+
+static inline int
+_Py_atomic_compare_exchange_ssize(Py_ssize_t *address, Py_ssize_t expected, Py_ssize_t value);
+
+static inline int
+_Py_atomic_compare_exchange_ptr(void *address, void *expected, void *value);
```
Comment (pitrou) [inline Include/cpython/pyatomic.h]: Should it be `void** address` for clarity?

File: Include/cpython/pyatomic.h [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+_Py_atomic_store_ptr_release(void *address, void *value);
+
+
+// Sequential consistency fence
+static inline void
+_Py_atomic_fence_seq_cst(void);
+
+// Release fence
+static inline void
+_Py_atomic_fence_release(void);
```
Comment (vstinner) [inline Include/cpython/pyatomic.h]:
```suggestion
// Sequential consistency fence
static inline void _Py_atomic_fence_seq_cst(void);

// Release fence
static inline void _Py_atomic_fence_release(void);
```

File: Include/cpython/pyatomic.h [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+
+static inline void
+_Py_atomic_store_ssize_relaxed(Py_ssize_t *address, Py_ssize_t value);
+
+// Stores `*address = value` (release operation)
+static inline void
+_Py_atomic_store_uint64_release(uint64_t *address, uint64_t value);
+
+static inline void
+_Py_atomic_store_ptr_release(void *address, void *value);
```
Comment (pitrou) [inline Include/cpython/pyatomic.h]: I'm not an expert, but why is it useful to expose "release" operations if no "acquire" operations are exposed?

File: Include/cpython/pyatomic_msc.h [hunk 14]
```diff
@@ -0,0 +1,860 @@
...
+
+// This is the implementation of Python atomic operations for MSVC if the
+// compiler does not support C11 or C++11 atomics.
+
+#include <intrin.h>
+
+
+static inline int
+_Py_atomic_add_int(int *address, int value)
+{
```
Comment (vstinner) [inline Include/cpython/pyatomic_msc.h:20]: Maybe add ``Py_BUILD_ASSERT(sizeof(int) == sizeof(long));``, since the cast is surprising.

Same remark for other casts below:

```
sizeof(uint8_t) == sizeof(char)
sizeof(uint16_t) == sizeof(short)
...
```

File: Include/cpython/pyatomic_msc.h [hunk 15]
```diff
@@ -0,0 +1,860 @@
...
+    return (int32_t)_InterlockedExchangeAdd((volatile long*)address, (long)value);
+}
+
+static inline int64_t
+_Py_atomic_add_int64(int64_t *address, int64_t value)
+{
+#if defined(_M_X64) || defined(_M_ARM64)
+    return (int64_t)_InterlockedExchangeAdd64((volatile __int64*)address, (__int64)value);
+#else
+    for (;;) {
```
Comment (vstinner) [inline Include/cpython/pyatomic_msc.h:53]: pycore_atomic.h uses a different style which may be more readable:

``do { ... } while (CMPXHG(...) == old_value);``

File: Include/cpython/pyatomic_msc.h [hunk 16]
```diff
@@ -0,0 +1,860 @@
...
+static inline uint32_t
+_Py_atomic_add_uint32(uint32_t *address, uint32_t value)
+{
+    return (uint32_t)_InterlockedExchangeAdd((volatile long*)address, (long)value);
+}
+
+static inline uint64_t
+_Py_atomic_add_uint64(uint64_t *address, uint64_t value)
+{
+    return (uint64_t)_Py_atomic_add_int64((volatile int64_t*)address, (int64_t)value);
```
Comment (vstinner) [inline Include/cpython/pyatomic_msc.h]: There is a compiler warning: ``different 'volatile' qualifiers``. _Py_atomic_add_int64() doesn't need volatile.

File: Include/cpython/pyatomic_msc.h [hunk 17]
```diff
@@ -0,0 +1,860 @@
...
+static inline uint32_t
+_Py_atomic_exchange_uint32(uint32_t *address, uint32_t value)
+{
+    return (uint32_t)_InterlockedExchange((volatile long*)address, (long)value);
+}
+
+static inline uint64_t
+_Py_atomic_exchange_uint64(uint64_t *address, uint64_t value)
+{
+    return (uint64_t)_Py_atomic_exchange_int64((volatile __int64*)address, (__int64)value);
```
Comment (vstinner) [inline Include/cpython/pyatomic_msc.h]: same compiler warning ``different 'volatile' qualifiers``, remove volatile? also cast to int64_t, not __int64.

File: Include/cpython/pyatomic_msc.h [hunk 18]
```diff
@@ -0,0 +1,860 @@
...
+static inline void
+_Py_atomic_store_uint32(uint32_t *address, uint32_t value)
+{
+    _InterlockedExchange((volatile long*)address, (long)value);
+}
+
+static inline void
+_Py_atomic_store_uint64(uint64_t *address, uint64_t value)
+{
+    _Py_atomic_exchange_int64((volatile __int64*)address, (__int64)value);
```
Comment (vstinner) [inline Include/cpython/pyatomic_msc.h]: same compiler warning: remove volatile?

File: Include/cpython/pyatomic.h [hunk 19]
```diff
@@ -0,0 +1,397 @@
...
+
+
+// Sequential consistency fence
```
Comment (vstinner) [inline Include/cpython/pyatomic.h]: I would prefer a more elaborated documentation, "fence" is kind of weak.

[__faststorefence doc](https://learn.microsoft.com/en-us/cpp/intrinsics/faststorefence?view=msvc-170):

> Guarantees that every previous memory reference, including both load and store memory references, is globally visible before any subsequent memory reference.

[_ReadWriteBarrier doc](https://learn.microsoft.com/en-us/cpp/intrinsics/readwritebarrier?view=msvc-170):

> Limits the compiler optimizations that can reorder memory accesses across the point of the call.

[dmb ish](https://devblogs.microsoft.com/oldnewthing/20220812-00/?p=106968):

> The data memory barrier ensures that all preceding writes are issued before any subsequent memory operations (including speculative memory access).

**Review by vstinner — CHANGES_REQUESTED** [review]
It's unclear to me if volatile must be used on the first parameter in the function definition, or not. For me, it's surprising to be able to cast ``int*`` to ``volatile long*`` in ``pyatomic_msc.h``. Maybe explain the black magic in the documentation at the top of  pyatomic.h?

I'm scared by the _ptr variant which takes a ``void*`` and is then casted to a pointer of a pointer (``void**``). It looks suspicious.

File: Include/cpython/pyatomic.h (hunk 9 above)
Comment (colesbury) [reply Include/cpython/pyatomic.h]: The motivation was two-fold: First, many of the _Py_atomic_compare_exchange calls in the nogil fork use constants as `expected` and this would be more verbose if it needs to be a pointer, e.g.:

```
_Py_atomic_compare_exchange_uint8(&m->v, LOCKED, UNLOCKED)
```

vs.

```
uint8_t expected = LOCKED:
_Py_atomic_compare_exchange_uint8(&m->v, &expected, UNLOCKED)
```

Second, I find this style (no pointer for expected) to be a bit less error-prone. I've been tripped up once or twice by having `expected` be modified when I didn't expect it.

I don't feel terribly strongly about this, so if there is a general preference for sticking closer to the C11-style API here, I can change it.

File: Include/cpython/pyatomic.h (hunk 10 above)
Comment (colesbury) [reply Include/cpython/pyatomic.h]: I expect to use at least one atomic operation on each of the data types here (but not every atomic op on every data type). I tried to be consistent on what's defined because it makes understanding what's available easier and testing easier.

File: Include/cpython/pyatomic.h (hunk 11 above)
Comment (colesbury) [reply Include/cpython/pyatomic.h]: The problem is that `void **` requires an explicit cast for almost every use, because things like `PyObject **` are not implicitly convertible to `void **`. In other words, currently we can write things like:

```
PyObject *old_exc = _Py_atomic_exchange_ptr(&tstate->async_exc, exc);
```

but if `address` was `void **address`, we'd have to write:

```
PyObject *old_exc = _Py_atomic_exchange_ptr((void **)&tstate->async_exc, exc);
```

File: Include/cpython/pyatomic.h (hunk 9 above)
Comment (pitrou) [reply Include/cpython/pyatomic.h]: AFAIU the main motivation for the C11 style APIs is for retry loops in lockless data structure implementations. A simplistic example (this may be embarassingly wrong):
```c

struct ListNode;

typedef struct ListNode {
  int value;
  struct ListNode* next;
} ListNode;

void ListAppend(ListNode* list, int new_value) {
  ListNode* new_node = (ListNode*) malloc(sizeof ListNode);
  new_node->value = new_value;
  new_node->next = NULL;
  ListNode* expected = NULL;
  while (_Py_atomic_compare_exchange_ptr(&list->next, &expected, new_node)) {
    list = expected;
    expected = NULL;
  }
}
```

File: Include/cpython/pyatomic.h (hunk 13 above)
Comment (colesbury) [reply Include/cpython/pyatomic.h]: You can always use stronger orderings for correctness (i.e., "seq_cst" everywhere instead of "acquire"). From a performance view, "release" is substantially faster than "seq_cst" stores on x86/x86-64, but "acquire" generates the same code as "seq_cst" loads on both x86/x86-64 and aarch64.

File: Include/cpython/pyatomic.h (hunk 6 above)
Comment (colesbury) [reply Include/cpython/pyatomic.h]: I slightly prefer *address* or *ptr* rather than *atomic*

File: Include/cpython/pyatomic.h (hunk 13 above)
Comment (pitrou) [reply Include/cpython/pyatomic.h]: I don't think CPython support is limited to x86 and ARM variants, so the set of atomic ops exposed should probably be made consistent nevertheless?

Also, using "seq_cst" in combination with "release" will probably make the code more difficult to reason about, than if "acquire" is exposed (and memory ordering is already hard to reason about!).

Comment (colesbury) [reply Include/cpython/pyatomic.h]: I'll add a `_Py_atomic_load_ptr_acquire` for consistency (and I think I can remove the `_Py_atomic_store_uint64_release`).

File: Include/cpython/pyatomic.h (hunk 11 above)
Comment (vstinner) [reply Include/cpython/pyatomic.h]: A large part of the Python C API uses macro to convert arguments to ``PyObject*``. Would it make sense to do the same here?

```
#define _Py_atomic_exchange_ptr(atomic, value) _Py_atomic_exchange_ptr(_Py_CAST(void**, atomic), (value))
```

Comment (colesbury) [reply Include/cpython/pyatomic.h]: I don't see the benefit of that style over the current approach, and it would silently allow passing some integer types to `_Py_atomic_exchange_ptr`.

File: Include/cpython/pyatomic_msc.h (hunk 15 above)
Comment (colesbury) [reply Include/cpython/pyatomic_msc.h:53]: This can be written as:

1. (current version) or equivalently with `while (1)`
```
    for (;;) {
        __int64 old_value = *(volatile __int64*)address;
        __int64 new_value = old_value + (__int64)value;
        if (old_value == _InterlockedCompareExchange64((volatile __int64*)address, new_value, old_value)) {
            return old_value;
        }
    }
```

2. (equivalent but with do {} while)
```
    do {
        __int64 old_value = *(volatile __int64*)address;
        __int64 new_value = old_value + (__int64)value;
        if (old_value == _InterlockedCompareExchange64((volatile __int64*)address, new_value, old_value)) {
            return old_value;
        }
    } while(1);
```

3. `do { ... } while (CMPXHG(...) == old_value);`
```
    __int64 old_value;
    do {
        old_value = *(volatile __int64*)address;
        __int64 new_value = old_value + (__int64)value;
    } while (old_value != _InterlockedCompareExchange64((volatile __int64*)address, new_value, old_value));
    return old_value;
```

I find (3) a bit harder to read and I like when the variables modified in the loop are scoped to the loop (as in 1 and 2), but if there is a strong preference for this style I will change it.

File: Include/cpython/pyatomic_msc.h (hunk 16 above)
Comment (colesbury) [reply Include/cpython/pyatomic_msc.h]: Thanks, I'll fix this.

File: Include/cpython/pyatomic.h (hunk 19 above)
Comment (colesbury) [reply Include/cpython/pyatomic.h]: The challenge I have is that it's really hard to describe what fences do in a way that's helpful and accurate. The above documentation is too strong for C11 fences.

There's https://en.cppreference.com/w/c/atomic/atomic_thread_fence, but I find it vague. And the C++ documentation (https://en.cppreference.com/w/cpp/atomic/atomic_thread_fence) is more detailed but really hard to understand.

File: Include/cpython/pyatomic_msc.h (hunk 15 above)
Comment (vstinner) [reply Include/cpython/pyatomic_msc.h:53]: I'm fine with the current style (1) ``for (;;)``.

**Comment by colesbury** [comment]
I have made the requested changes; please review again.

File: Include/cpython/pyatomic.h (hunk 19 above)
Comment (vstinner) [reply Include/cpython/pyatomic.h]: Please add one or two of these links here. It's ok to have references to external doc, it's better than no doc :-)

**Review by vstinner — COMMENTED** [review]
The glib library provides g_atomic_int_dec_and_test() to implement reference counting. How would you reimplement it with your API? I'm not used to atomic variables and I never know how to use them *correctly*.

**Comment by colesbury** [comment]
> The glib library provides g_atomic_int_dec_and_test() to implement reference counting. How would you reimplement it with your API?

The decrement would look like the following. The `atomic_add` functions return the previous value like the C11/C++11 `atomic_fetch_add` function.

```c
if (_Py_atomic_add_ssize(&op->ob_refcnt, -1) == 1) {
  // refcnt is zero, dealloc
}
```

**Comment by colesbury** [comment]
@vstinner, would you please look this over again when you have a chance?

File: PCbuild/pythoncore.vcxproj.filters [hunk 20]
```diff
@@ -423,6 +423,12 @@
     <ClInclude Include="..\Include\cpython\pyarena.h">
       <Filter>Include</Filter>
     </ClInclude>
+    <ClInclude Include="..\Include\cpython\pyatomic.h">
+      <Filter>Include</Filter>
+    </ClInclude>
+    <ClInclude Include="..\Include\cpython\pyatomic_msc.h">
+      <Filter>Include</Filter>
+    </ClInclude>
```
Comment (vstinner) [inline PCbuild/pythoncore.vcxproj.filters:434]: You should add GCC here:

```suggestion
    <ClInclude Include="..\Include\cpython\pyatomic_gcc.h">
      <Filter>Include</Filter>
    </ClInclude>
    <ClInclude Include="..\Include\cpython\pyatomic_msc.h">
      <Filter>Include</Filter>
    </ClInclude>
```

It's just for the UI, not to build Python.

File: Include/cpython/pyatomic.h (hunk 6 above)
Comment (vstinner) [reply Include/cpython/pyatomic.h]: * https://en.cppreference.com/w/c/atomic uses ``obj`` name.
* https://www.ibm.com/docs/en/zos/2.1.0?topic=c11-atomic-load also uses *obj* name (C11 functions)
* pycore_atomic.h uses ``ATOMIC_VAL`` name
* pycore_atomic_func.h uses ``var`` name

I dislike the ``address`` name. What you pass is not a ``void*`` pointer or an ``uintptr_t`` address, but a reference to an atomic variable. I suggest to use *atomic* or *obj* name.

For add operation, the C11 API uses *arg* for the second parameter name. I'm fine with value.

File: Include/cpython/pyatomic.h (hunk 9 above)
Comment (vstinner) [reply Include/cpython/pyatomic.h]: If the second argument is a reference (``&expected``), does it mean that it changes the value of the second argument (``*expected``)?

For C11 atomic_exchange(), the second argument is not a pointer, but a value (integer), no? https://en.cppreference.com/w/c/atomic/atomic_exchange 

C11 atomic_compare_exchange_strong() and atomic_compare_exchange_weak() use a pointer for expected. But this API writes into ``*expected`` if the ``*obj`` is not equal to ``*expected``.

> The behavior of atomic_compare_exchange_* family is as if the following was executed atomically:

```c
if (memcmp(obj, expected, sizeof *obj) == 0) {
    memcpy(obj, &desired, sizeof *obj);
    return true;
} else {
    memcpy(expected, obj, sizeof *obj);
    return false;
}
```

For this header fie, I would prefer to not have two flavors, the API is already quite long! I would prefer to have a single flavor. If there is an usecase where setting expected is relevant, I suggest to use a pointer for the second argument.

In short, I agree to change the API to ``int _Py_atomic_compare_exchange_int32(int32_t *obj, int32_t *expected, int32_t desired)``. The behavior should be well documented.

*obj*, *expected* and *desired* names come from the C11 API.

File: Include/cpython/pyatomic.h (hunk 11 above)
Comment (vstinner) [reply Include/cpython/pyatomic.h]: Right, in that case I'm fine with the surprising ``void*`` type.

**Review by vstinner — COMMENTED** [review]
We should decide _Py_atomic_compare_exchange() second argument should be a pointer or not. Apparently, a pointer covers more cases and so should be used.

File: Include/cpython/pyatomic.h (hunk 6 above)
Comment (colesbury) [reply Include/cpython/pyatomic.h]: I don't think `obj` makes as much sense in this API because, unlike the C11 API, they are not pointers to atomic objects (e.g., `_Atomic type`). They are just pointers to plain C variables. I think that's an important distinction.

If you dislike `address`, I can change it to `ptr`, which we commonly use and is a bit more concise.

File: Include/cpython/pyatomic.h (hunk 9 above)
Comment (colesbury) [reply Include/cpython/pyatomic.h]: I'll make the second argument a pointer like the C11 API.

File: Include/cpython/pyatomic.h (hunk 6 above)
Comment (colesbury) [reply Include/cpython/pyatomic.h]: For context, the GCC documentation for built-in atomic functions uses `ptr`. This is the API that `pyatomic.h` most closely matches.

https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html

**Comment by colesbury** [comment]
@vstinner, I've made the second argument to `_Py_atomic_compare_exchange_` functions a pointer. I renamed the first argument to all the atomic functions to `ptr` for the reasons in my above comment. I can change it to `obj` if you prefer.

File: Include/cpython/pyatomic.h [hunk 21]
```diff
@@ -0,0 +1,406 @@
...
+_Py_atomic_add_uintptr(uintptr_t *ptr, uintptr_t value);
+
+static inline Py_ssize_t
+_Py_atomic_add_ssize(Py_ssize_t *ptr, Py_ssize_t value);
+
+// Performs an atomic compare-and-exchange. If `*ptr` and `*expected` are
+// equal, then `desired` is stored in `*ptr`. Otherwise `*expected` is updated
+// with the current value of `*ptr`. Returns 1 on success and 0 on failure.
+// These correspond to the "strong" variations of the C11
+// atomic_compare_exchange_* functions.
```
Comment (vstinner) [inline Include/cpython/pyatomic.h]: There is a single function, no?

```suggestion
// These correspond to the C11
// atomic_compare_exchange_strong() function.
```

atomic_compare_exchange_weak_explicit() has 2 additional arguments (of type memory_order).

**Comment by vstinner** [comment]
> I renamed the first argument to all the atomic functions to ptr for the reasons in my above comment. I can change it to obj if you prefer.

Please rename the argument to ``obj``.

---

Naming is a hard problem :-( If I want to write documentation for that, I would have issues to explain why I have to pass a "pointer" (*ptr*) to these functions.

https://en.cppreference.com/w/c/atomic/atomic_load says:

> Parameters
> obj 	- 	pointer to the atomic object to access
> order 	- 	the memory synchronization ordering for this operation 

The first argument is a **pointer** to the atomic **object** to access.

It's right that it's a pointer, so "atomic_ptr", "atomic_obj_ptr", "patomic", ... names would be good. But to make the name shorter, I would suggest to omit "pointer", and so just say "object" (or "atomic", but you dislike this name, so let's skip it).

I would prefer to document that _Py_atomic_load_int32(obj) gets the value of the atomic **object** *obj*, rather than the value of the atomic **pointer** *ptr*.

**Comment by pitrou** [comment]
"object" in a CPython context is really misleading. I also don't understand what the issue with "pointer" or "ptr" is.

**Comment by pitrou** [comment]
As you prefer... This is a private low-level API anyway, so the name of arguments is hardly fundamental.

**Comment by colesbury** [comment]
I've renamed `ptr` to `obj`

**Comment by vstinner** [comment]
Well. I have "a few more remarks", but I decided to copy this PR and make my changes directly there: please see my PR #108701.

**Comment by vstinner** [comment]
I merged https://github.com/python/cpython/pull/108701 which includes my coding style changes. Thanks!
